File: frontend/src/lib/parseQuestion.ts
export interface ParsedQuestion {
  body: string;
  options: string[];
}

export const OPTION_MARKERS = ["①", "②", "③", "④"];

/**
 * 본문에서 ①②③④ 옵션을 분리한다.
 * - 옵션 마커 등장 이전 줄은 모두 본문(body)
 * - 옵션 마커로 시작하는 줄은 options 배열에 적재 (마커 제거)
 * - 옵션 마커로 시작하지 않지만 이미 옵션 수집이 시작됐다면 직전 옵션에 이어붙임 (들여쓰기 보존)
 * - 본문은 마크다운 그대로 보존하여 호출처에서 react-markdown으로 렌더한다.
 */
export function parseQuestion(content: string): ParsedQuestion {
  const lines = content.split("\n");
  const bodyLines: string[] = [];
  const options: string[] = [];

  for (const line of lines) {
    const trimmed = line.trim();
    if (OPTION_MARKERS.some((m) => trimmed.startsWith(m))) {
      options.push(trimmed.replace(/^[①②③④]\s*/, ""));
    } else if (options.length === 0) {
      bodyLines.push(line);
    } else {
      // 다중 라인 옵션: 들여쓰기·줄바꿈 보존하고 직전 옵션에 이어붙인다.
      // 정처기 필기처럼 보기 안에 코드 조각이 들어가는 케이스에 필요.
      options[options.length - 1] += "\n" + line;
    }
  }

  return { body: bodyLines.join("\n"), options };
}

File: ios/Sqldpass/Core/Content/ParseQuestion.swift
import Foundation

/// 백엔드 응답의 `content` 한 덩어리를 본문(body) / 보기(options) 로 분리.
///
/// Android `mobile/.../text/QuestionParser.kt` (`parseQuestion`) 와 1:1 동치.
/// ①②③④ 마커로 시작하는 줄을 option, 그 이전은 body, option 시작 후 마커 없는 줄은 직전 option 에 이어붙임.
struct ParsedQuestion {
    let body: String
    let options: [String]
}

enum QuestionParser {
    private static let optionMarkers: [Character] = ["①", "②", "③", "④"]
    private static let leadingMarkerRegex: NSRegularExpression = {
        try! NSRegularExpression(pattern: "^[①②③④]\\s*")
    }()

    static func parse(_ content: String) -> ParsedQuestion {
        let lines = content.components(separatedBy: "\n")
        var bodyLines: [String] = []
        var options: [String] = []
        for line in lines {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if let first = trimmed.first, optionMarkers.contains(first) {
                let ns = trimmed as NSString
                let stripped = leadingMarkerRegex.stringByReplacingMatches(
                    in: trimmed,
                    options: [],
                    range: NSRange(location: 0, length: ns.length),
                    withTemplate: ""
                )
                options.append(stripped)
            } else if options.isEmpty {
                bodyLines.append(line)
            } else {
                let last = options.removeLast()
                options.append(last + "\n" + line)
            }
        }
        return ParsedQuestion(
            body: bodyLines.joined(separator: "\n"),
            options: options
        )
    }
}
